ui/assets/js/common.test.js

```diff
@@ -19,3 +19,90 @@ describe("ExpandVars", () => {
     });
 });
+
+describe("UrlToRepo", () => {
+    test("Generate url from repo with default values", () => {
+        const repo = {
+            url: "https://www.github.com/YourOrganization/RepoOne.git",
+            "url-pattern":
+            {
+                "base-url": "{url}/blob/{rev}/{path}{anchor}",
+                anchor: "#L{line}"
+            }
+        };
+        const path = "test.txt"
+        const line = null
+        const rev = "main"
+        expect(UrlToRepo(repo, path, line, rev)).toBe(
+            "https://www.github.com/YourOrganization/RepoOne/blob/main/test.txt"
+        );
+    });
+
+    test("Generate url from repo with default values and line", () => {
+        const repo = {
+            url: "https://www.github.com/YourOrganization/RepoOne.git",
+            "url-pattern":
+            {
+                "base-url": "{url}/blob/{rev}/{path}{anchor}",
+                anchor: "#L{line}"
+            }
+        };
+        const path = "test.txt"
+        const line = "12"
+        const rev = "main"
+        expect(UrlToRepo(repo, path, line, rev)).toBe(
+            "https://www.github.com/YourOrganization/RepoOne/blob/main/test.txt#L12"
+        );
+    });
+
+    test("Generate url from for ssh style repo with default values", () => {
+        const repo = {
+            url: "git@github.com:YourOrganization/RepoOne.git",
+            "url-pattern":
+            {
+                "base-url": "{url}/blob/{rev}/{path}{anchor}",
+                anchor: "#L{line}"
+            }
+        };
+        const path = "test.txt"
+        const line = null
+        const rev = "main"
+        expect(UrlToRepo(repo, path, line, rev)).toBe(
+            "//github.com/YourOrganization/RepoOne/blob/main/test.txt"
+        );
+    });
+
+    test("Generate url from for ssh bitbucket mercurial style repo", () => {
+        const repo = {
+            url: "ssh://hg@bitbucket.org/YourOrganization/RepoOne",
+            "url-pattern":
+            {
+                "base-url" : "{url}/src/main/{path}{anchor}",
+                "anchor" : "#{filename}-{line}"
+            }
+        };
+        const path = "test.txt"
+        const line = null
+        const rev = "main"
+        expect(UrlToRepo(repo, path, line, rev)).toBe(
+            "//bitbucket.org/YourOrganization/RepoOne/src/main/test.txt"
+        );
+    });
+
+    test("Generate url from for ssh bitbucket style repo with port", () => {
+        const repo = {
+            url: "ssh://git@bitbucket.org:7999/YourOrganization/RepoOne",
+            "url-pattern":
+            {
+                "base-url" : "{url}/src/main/{path}{anchor}",
+                "anchor" : "#{filename}-{line}"
+            }
+        };
+        const path = "test.txt"
+        const line = null
+        const rev = "main"
+        expect(UrlToRepo(repo, path, line, rev)).toBe(
+            "//bitbucket.org/YourOrganization/RepoOne/src/main/test.txt"
```

Comment:
Should this repo URL contain the port? i.e. would we expect this to be `"//bitbucket.org:7999/YourOrganization/RepoOne/src/main/test.txt"` in all cases?

Reply:
Thanks for the reviewing this.

My actual goal is to get Hound working with BitBucket Server. From https://confluence.atlassian.com/bitbucketserver074/enabling-ssh-access-to-git-repositories-in-bitbucket-server-1019386434.html#EnablingSSHaccesstoGitrepositoriesinBitbucketServer-SSHbaseURL
```
Resulting SSH URL for a repo: ssh://git@bitbucket.atlassian.com:7999/<projectname>/<reponame>.git
Bitbucket Server base URL: https://bitbucket.atlassian.com
```
In this case the port is required in the config.json when specifying the ssh url, but the port needs to be dropped when resolving the url when clicking through to a file after search.

I guess the best approach would be to allow specifying a `{port}` in `base-url` to allow the user to pick whether they want the port or not.

As an aside, I thought this was the only change needed to get it working with BitBucket Server, but after finally getting around to building and running my branch I've noticed that the required URL is in the form:
https://bitbucket.atlassian.com/projects/<projectname>/repos/<reponame>/browse which would require further additions to the `UrlToRepo` method. This diversion may not be what you want, so won't open a PR unless you think it's worthwhile.

Reply:
No stress. Thanks for getting back to me.

> leave the port in by default, and then to make a special exemption for bitbucket / atlassian-hosted repositories.
Sounds good to me.

> I'd also be open to a PR to handle Bitbucket URLs, either as part of this PR or in a separate one.

Ok cool thanks - I'm not really sure where the best place to add that logic would be. I'll update this PR so that the ports are not ommited, and then you can add the exemption for on prem bitbucket repositories (or give me some guidance on what you have in mind and I can do it).

Reply:
For sure. I think in that function itself, there's already some edge cases being handled (like [this one for Github wikis](https://github.com/hound-search/hound/blob/main/ui/assets/js/common.js#L15). I wouldn't be mad at exceptional handling of some sort for Bitbucket in there as well. If that method gets too unruly, it shouldn't be too difficult to refactor later, too.

Reply:
Added full support for the Bitbucket server style repos by adding expansion vars. Tested it against our deployed bitbucket and click through to source is working now. Everything else should have remained backwards compatible.